### DISCUSSION TITLE
multiple databases

### DISCUSSION THREAD

**Discussion by elsimqoz** [fastapi/fastapi#6941] [Questions] [answered]
Hi. Are there any way to use multiple databases with FastApi like [flask-sqlalchemy](https://flask-sqlalchemy.palletsprojects.com/en/2.x/binds)?
I have founded [SO question](https://stackoverflow.com/questions/64190938/fastapi-to-handle-multiple-db-connections-via-sqlalchemy) about it, but this question is without any comments. Of course I can create [several dependencies](https://fastapi.tiangolo.com/tutorial/sql-databases/#create-a-dependency), but it doesn't seem like correct way. So what shall I do? Thanks.

**Comment by n8sty** [comment]
Creating multiple dependencies seems like a perfectly reasonable method to achieve using multiple databases in your application. I've done this in a few different applications. I wouldn't do what's being attempted in the SO question you link to, just use your dependencies explicitly where you need them.

**Comment by elsimqoz** [comment]
@n8sty firstly, thanks for your answer.
Secondly, how do you use multiple databases in your project? I prefer to use with middleware and it seems it's impossible to do it in this case.
```python
from fastapi import FastAPI

from fastapi_sqlalchemy import DBSessionMiddleware


app = FastAPI()
app.add_middleware(middleware_class=DBSessionMiddleware, db_url=...)
```

**Comment by Kludex** [comment]
`DBSessionMiddleware` makes the session available through a global `db` object. If that was not the case, it would have been possible. 

If you want to stick with this solution, you can think of creating your own middleware, a "FactoryDBSessionMiddleware" :kissing:

**Comment by n8sty** [comment]
It's project specific, but one example is the case where you want to write to a "writer" database endpoint but read from a "reader". In that case, I'll create two dependencies, `db_writer` and `db_reader` and build up more complicated dependencies from there. I push a lot of logic into dependencies, so if I had a `/posts` route, I'd then have a `create_posts` dependency, that itself depends on `db_writer, and something similar for the `GET` route. This is a bit verbose, but it's quite explicit.

I've tried using middleware and I opted for the above approach because it removes having to push logic into the middleware that effectively chooses what database to use. If you're only ever using two instances of the same database, let's say two Postgres databases in the reader/writer setup I mentioned, then middleware is quite slick. But, if you want to use different source types, the middleware can get bloated. Obviously, all just my opinion.

**Comment by elsimqoz** [comment]
Thanks

**Comment by Giovanni-94** [comment]
Any examples?

**Comment by elsimqoz** [comment, marked as answer]
@Giovanni-94 smth like
```Python
from sqlalchemy.ext.declarative import as_declarative

@as_declarative()
class ModelDB1:
    pass

@as_declarative()
class ModelDB2:
    pass

###

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

engine1 = create_engine('postgresql://dev@postgres:5432/db1')
engine2 = create_engine('postgresql://dev@postgres:5432/db2')

DBSession = sessionmaker(...)
DBSession.configure(binds={ModelDB1: engine1, ModelDB2: engine2})

def get_session():
    db = DBSession()
    try:
        yield db
    finally:
        db.close()

###

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

router = APIRouter()

@router.get('/')
def get(self, session: Session = Depends(get_session)):
    pass
```

**Comment by Giovanni-94** [comment]
thanks @elsimqoz for the reply, any example with tortoise orm?

**Comment by Allan-Nava** [comment]
> thanks @elsimqoz for the reply, any example with tortoise orm?

Me too i need it!!

**Comment by IvanMSP** [comment]
hello, with this implementation, I'm get this error:
`sqlalchemy.exc.UnboundExecutionError: Could not locate a bind configured on SQL expression or this Session.`
Could you help me, please

**Comment by thinkjrs** [comment]
> hello, with this implementation, I'm get this error: `sqlalchemy.exc.UnboundExecutionError: Could not locate a bind configured on SQL expression or this Session.` Could you help me, please

I'm not sure what your specific issue is, however, it seems more an SQLAlchemy issue than Fastapi. 

[Here](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.sessionmaker) is the section on the `sessionmaker` factory in the SQLAlchemy documentation. 

Generally, you want to make sure you call the `configure` method above _before_ you instantiate and use your session object (e.g. `DBSession` above).

**Comment by FatPuffer** [comment]
I use tortoise to configure read-write separation, but in the actual use process, there is a problem and I need help
```py
from typing import Type

from tortoise import Model


class Router:
    @staticmethod
    def db_for_read(model: Type[Model]):
        return "slave"

    @staticmethod
    def db_for_write(model: Type[Model]):
        return "master"


TORTOISE_ORM = {
    'connections': {
        'master': {
            # 'engine': 'tortoise.backends.asyncpg',
            'engine': 'tortoise.backends.mysql',
            # 'engine': 'tortoise.backends.aiomysql',
            'credentials': {
                'host': '127.0.0.1',
                'port': '3306',
                'user': 'root',
                'password': 'password',
                'database': 'base',
                'minsize': 1,
                'maxsize': 5,
                'charset': 'utf8mb4',
                'echo': True
            }
        },
        'slave': {
            'engine': 'tortoise.backends.mysql',
            'credentials': {
                'host': '127.0.0.1',
                'port': '3306',
                'user': 'root',
                'password': 'password',
                'database': 'base',
                'minsize': 1,
                'maxsize': 5,
                'charset': 'utf8mb4',
                'echo': True
            }
        },
    },
    'apps': {
        'models': {
            'models': ['apps.users.models', "aerich.models"],
            # 'models': {"models": ["app.models.model"]},
            'default_connection': 'master',
        }
    },
    'routers': [Router]
}


>>> admin = AdminModel.get_or_none(name='admin')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\fatpuffer\AppData\Local\pypoetry\Cache\virtualenvs\baseapi-FOv_RZhh-py3.7\lib\site-packages\tortoise\models.py", line 1343, in get_or_none
    db = using_db or cls._choose_db()
  File "C:\Users\fatpuffer\AppData\Local\pypoetry\Cache\virtualenvs\baseapi-FOv_RZhh-py3.7\lib\site-packages\tortoise\models.py", line 1031, in _choose_db
    db = router.db_for_read(cls)
  File "C:\Users\fatpuffer\AppData\Local\pypoetry\Cache\virtualenvs\baseapi-FOv_RZhh-py3.7\lib\site-packages\tortoise\router.py", line 36, in db_for_read
    return self._db_route(model, "db_for_read")
  File "C:\Users\fatpuffer\AppData\Local\pypoetry\Cache\virtualenvs\baseapi-FOv_RZhh-py3.7\lib\site-packages\tortoise\router.py", line 31, in _db_route
    return connections.get(self._router_func(model, action))
  File "C:\Users\fatpuffer\AppData\Local\pypoetry\Cache\virtualenvs\baseapi-FOv_RZhh-py3.7\lib\site-packages\tortoise\router.py", line 18, in _router_func
    for r in self._routers:
TypeError: 'NoneType' object is not iterable
>>>
```

**Comment by verigle** [comment]
does 

> @Giovanni-94 smth like
> 
> ```python
> from sqlalchemy.ext.declarative import as_declarative
> 
> @as_declarative()
> class ModelDB1:
>     pass
> 
> @as_declarative()
> class ModelDB2:
>     pass
> 
> ###
> 
> from sqlalchemy import create_engine
> from sqlalchemy.orm import sessionmaker
> 
> engine1 = create_engine('postgresql://dev@postgres:5432/db1')
> engine2 = create_engine('postgresql://dev@postgres:5432/db2')
> 
> DBSession = sessionmaker(...)
> DBSession.configure(binds={ModelDB1: engine1, ModelDB2: engine2})
> 
> def get_session():
>     db = DBSession()
>     try:
>         yield db
>     finally:
>         db.close()
> 
> ###
> 
> from fastapi import APIRouter, Depends
> from sqlalchemy.orm import Session
> 
> router = APIRouter()
> 
> @router.get('/')
> def get(self, session: Session = Depends(get_session)):
>     pass
> ```

does it support read from one database and writer to another database ?

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by codebydant** [comment]
> hello, with this implementation, I'm get this error: `sqlalchemy.exc.UnboundExecutionError: Could not locate a bind configured on SQL expression or this Session.` Could you help me, please


I am getting that same issue, @IvanMSP did you manage to solve it? 

The approach above it works for me when running standard queries directly to the Table object such as:

```python
@router.get(
    "/data",
    status_code=status.HTTP_200_OK,
    summary="Get all data",
)
async def get_data_all(db: Session = Depends(get_db)) :

    response = db.query(MYTABLE).all()

    if response is not None:
        return jsonable_encoder(response)
    raise HTTPException(status_code=411, ...)
```

but, when running a SQL RAW query it fails:

```python

@router.post("/sql")
async def run_sql_query(db: Session = Depends(get_db)):
    query = text(request.sql)

    try:
        response = db.execute(query).fetchall()
    except Exception as e:
        print("ERROR: ", e)
    else:
        return response
```

then I got:

`Could not locate a bind configured on SQL expression or this Session.`

any ideas? I know this is a more sqlalchemy issue, but any help would be appreciated.

## possible issues
I think it may be related that each "bind" is associated with each model in the database and the raw SQL query is not bounded to a specific table or engine connection, so the bind connection is not generated by a raw SQL query and that's why fails

**Reply by robbertvdh** [reply]
@dtcMLOps I am running into the exact same issue, and came to the same conclusion regarding the bind connection. Did you manage to find a solution for running raw SQL queries?

**Reply by codebydant** [reply]
Hi @robbertvdh I have avoided using RAW SQL because SQL injection might be a security vulnerability.
